[Infra] Disable infra plugin in serverless projects

crespocarlos opened the pull request:
closes https://github.com/elastic/kibana/issues/158925

## Summary

This PR disables the infra plugin in the `oblt` and `security` projects

### Oblt
**Navbar**
<img width="300" alt="image" src="https://github.com/elastic/kibana/assets/2767137/f88fd2a5-ed58-4bcf-b5ad-5b25357daf5c">

**Saved Objects**
<img width="800" alt="image" src="https://github.com/elastic/kibana/assets/2767137/4140760c-2572-4391-8977-bc97fcd14bb8">



**Server-side APIs**
<img width="800" alt="image" src="https://github.com/elastic/kibana/assets/2767137/eb22b024-dfd5-4c7f-b760-bfd2c633f759">

**Infra routes**
/metrics/inventory

<img width="500" alt="image" src="https://github.com/elastic/kibana/assets/2767137/6ddba766-c6f0-40d0-9183-2bcbae348d52">

**Alert rules**
<img width="300" alt="image" src="https://github.com/elastic/kibana/assets/2767137/d03d087e-d562-425b-a4d9-a17d5fe0b158">


### APM

Infra links hidden in the Transaction "Investigate" pop up menu
<img width="800" alt="image" src="https://github.com/elastic/kibana/assets/2767137/d9364d55-6e06-4ac0-8668-6ce2331a6250">

Infra links removed from Services Overview Instance actions
<img width="800" alt="image" src="https://github.com/elastic/kibana/assets/2767137/73410e5c-be6c-4174-8df0-76fb6a25b2f6">

** Server Routes **

<img width="800" alt="image" src="https://github.com/elastic/kibana/assets/2767137/488f639b-81e8-497d-a55e-4e7f31644bc5">

<img width="800" alt="image" src="https://github.com/elastic/kibana/assets/2767137/4366005a-2eb8-4f33-8d1e-09865cad06b4">

<img width="800" alt="image" src="https://github.com/elastic/kibana/assets/2767137/46b34d0d-8171-4d8d-bfb8-fd4c32583fbb">





### Security

**Saved Objects**
<img width="800" alt="image" src="https://github.com/elastic/kibana/assets/2767137/14f08e7e-565b-42e9-8fc4-753d1b54c14a">



**Server-side APIs**
<img width="800" alt="image" src="https://github.com/elastic/kibana/assets/2767137/eb22b024-dfd5-4c7f-b760-bfd2c633f759">

**Infra routes**
/metrics/inventory

<img width="500" alt="image" src="https://github.com/elastic/kibana/assets/2767137/6ddba766-c6f0-40d0-9183-2bcbae348d52">


### How to test

- Start a local ES instance
- run `yarn serverless-oblt` and `yarn serverless-security`
- Validate the screenshots above.

crespocarlos committed:
Disable infra in serverless

apmmachine commented:
### :robot: GitHub comments

<details><summary>Expand to view the GitHub comments</summary>
<p>

Just comment with:
- `/oblt-deploy` : Deploy a Kibana instance using the Observability test environments.
- `/oblt-deploy-serverless` : Deploy a serverless Kibana instance using the Observability test environments.
- `run` `elasticsearch-ci/docs` : Re-trigger the docs validation. (use unformatted text in the comment!)

</p>
</details>

crespocarlos committed:
Merge branch 'main' of github.com:elastic/kibana into 158925-disable-infra-in-serverless-projets

crespocarlos commented:
@elasticmachine merge upstream

kibanamachine committed:
Merge branch 'main' into 158925-disable-infra-in-serverless-projets

crespocarlos commented:
buildkite test this

crespocarlos commented:
@elasticmachine merge upstream

kibanamachine committed:
Merge branch 'main' into 158925-disable-infra-in-serverless-projets

crespocarlos commented:
@elasticmachine merge upstream

kibanamachine committed:
Merge branch 'main' into 158925-disable-infra-in-serverless-projets

weltenwort commented:
/oblt-deploy-serverless

crespocarlos committed:
Turn infra into optional dependency in APM

crespocarlos commented:
/oblt-deploy-serverless

crespocarlos commented:
@elasticmachine merge upstream

kibanamachine committed:
Merge branch 'main' into 158925-disable-infra-in-serverless-projets

crespocarlos commented:
@elasticmachine merge upstream

kibanamachine committed:
Merge branch 'main' into 158925-disable-infra-in-serverless-projets

crespocarlos commented:
@elasticmachine merge upstream

kibanamachine committed:
Merge branch 'main' into 158925-disable-infra-in-serverless-projets

kibanamachine committed:
Merge branch 'main' into 158925-disable-infra-in-serverless-projets

elasticmachine commented:
Pinging @elastic/infra-monitoring-ui (Team:Infra Monitoring UI)

elasticmachine commented:
Pinging @elastic/apm-ui (Team:APM)

weltenwort reviewed (commented):
After struggling a bit with deploying a serverless dev env, this seems to work as expected. 

Two things came to mind, though:

1. Can we also mark the `infra` plugin as optional in the apm plugin's server-side types? That way we can let the compiler enforce the softness of the dependency in the server-side code too.
2. When starting Kibana I noticed that it disables the plugin `ux` because it has a dependency on `infra`. Can we clarify whether `ux` is supposed to be disabled?

crespocarlos committed:
Make infra plugin optional in APM server; Hide links to Infra in the UI

elasticmachine commented:
Pinging @elastic/uptime (Team:uptime)

crespocarlos commented:
Hi @weltenwort , thanks for reviewing it.

> 1. Can we also mark the `infra` plugin as optional in the apm plugin's server-side types? That way we can let the compiler enforce the softness of the dependency in the server-side code too.

Done! I had to touch 3 endpoints. While doing this I also found a couple more places in the UI that depended on `InfraLocators`. Fixed those too.

> 2. When starting Kibana I noticed that it disables the plugin `ux` because it has a dependency on `infra`. Can we clarify whether `ux` is supposed to be disabled?

Apparently `ux` didn't even need to depend on `infra`. I've removed it from the dependency list.

crespocarlos commented:
@elasticmachine merge upstream

kibanamachine committed:
Merge branch 'main' into 158925-disable-infra-in-serverless-projets

crespocarlos commented:
@elasticmachine merge upstream

kibanamachine committed:
Merge branch 'main' into 158925-disable-infra-in-serverless-projets

tonyghiani commented:
I closed it by mistake, I was in the wrong browser tab I'm sorry 🙏

crespocarlos commented:
@elasticmachine merge upstream

kibanamachine committed:
Merge branch 'main' into 158925-disable-infra-in-serverless-projets

weltenwort reviewed (approved):
LGTM, tested on a QA serverless project :+1: thanks for taking care of this.

I'll leave it to the apm-ui team to judge whether throwing exceptions is the appropriate way of handling a disabled `infra` dependency.

MiriamAparicio reviewed (approved):
LGTM

crespocarlos commented:
@elasticmachine merge upstream

kibanamachine committed:
Merge branch 'main' into 158925-disable-infra-in-serverless-projets

crespocarlos commented:
@elasticmachine merge upstream

kibanamachine committed:
Merge branch 'main' into 158925-disable-infra-in-serverless-projets
